utils/s2n_safety.c

```diff
@@ -78,6 +78,49 @@ bool s2n_constant_time_equals(const uint8_t *a, const uint8_t *b, const uint32_t
     return (xor == 0);
 }
 
+/*  See specification of this function in s2n_safety.h */
+bool s2n_constant_time_equals_partial(const uint8_t *const a,
+                                      const uint8_t *const b,
+                                      const uint32_t len)
+{
+    bool arrays_equal = true;
+    /* iterate over each byte in the slices */
+    for (uint32_t i = 0; i < len; i++)
+    CONTRACT_ASSIGNS(i, arrays_equal)
+    CONTRACT_INVARIANT(i <= len)
+    CONTRACT_INVARIANT(arrays_equal ==
+                       __CPROVER_forall { unsigned j; (j >= 0 && j < i) ==> (a[j] == b[j]) })
+    CONTRACT_DECREASES(len - i)
+    {
+        arrays_equal = arrays_equal && (a[i] == b[i]);
+    }
+
+    /* Substiture i = len into the loop invariant to get... */
+    CONTRACT_ASSERT(arrays_equal ==
+                     __CPROVER_forall { unsigned j; (j >= 0 && j < len) ==> (a[j] == b[j]) },
+                     "Post-loop assertion");
+    return arrays_equal;
+}
+
+
+/*  See specification of this function in s2n_safety.h */
+bool s2n_constant_time_equals_total(const uint8_t *const a,
+                                    const uint8_t *const b,
+                                    const uint32_t len)
+{
+    if (len == 0) {
+        return true;
+    }
+
+    if (a != NULL && b != NULL) {
```

Comment:
Is there a reason we can't do?

     if (a == NULL || b == NULL) {
         return false;
     }
     
     return s2n_constant_time_equals_partial(a, b, len);
     
that's just more in line with s2n's style and since this is a core safety function it's nice to be a strong example.

Reply:
Yes... it could be done that way. My personal programming style is for a function's CFG to have a single-entry and single-exit point, and therefore avoid "early returns" like this. But... always happy to adopt the style of an existing project.

Reply:
I'm not saying there's only one right way, there isn't, but the reason we do it this way in s2n is to keep average branch depth as shallow as possible. For a case like this it doesn't matter much, but when you're a few calls deep it creates combinatorial complexity and it's a mental tax on the conditions and state that a reader has to keep in their head. Consistency is then the reason to do it even in the negligible cases.

Reply:
indeed - consistency and readability have priority.

Reply:
Done